### PR TITLE
feat: MLS Client

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Introduce `MLSClient` which is a KMM wrapper around https://github.com/wireapp/core-crypto

This will enable us to register MLS clients and communicate using the [messaging layer protocol](https://messaginglayersecurity.rocks)

### Testing

Unit tests

### Notes

- Currently only JVM and Android targets have functional implementations.
- Enabled  `isCoreLibraryDesugaringEnabled` because CoreCrypto `uses java.time.Duration` which is only available since API 26
- CoreCrypto is still in development so the API will change in the future (I left some TODOs) 

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: cryptography/src/commonJvmAndroid/kotlin/com.wire.kalium.cryptography/MLSClientImpl.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import com.wire.crypto.Invitee
+import java.time.Duration
+
+@OptIn(ExperimentalUnsignedTypes::class)
+actual class MLSClientImpl actual constructor(
+    rootDir: String,
+    databaseKey: String,
+    clientId: CryptoQualifiedClientId) : MLSClient {
+
+    private var coreCrypto: CoreCrypto
```
Comment (MohamadJaara) [inline cryptography/src/commonJvmAndroid/kotlin/com.wire.kalium.cryptography/MLSClientImpl.kt:15]: Suggestion to make coreCrypto immutable 
```suggestion
    private val coreCrypto: CoreCrypto
```

**Comment by typfel** [comment]
Tests were initially failing on Android with:
```
Installation failed with message Error: android.os.ParcelableException: java.io.IOException: Requested internal only, but not enough space.
```

Probably because the MLS library was too big since the native binaries included debug symbols. I re-published the MLS library this time build for release, but the tests were still failing. I then disabled the gradle build cache and then the tests passed.

tl;dr: I'll revert my attempt to fix the tests.

File: .github/workflows/gradle-android-tests.yml [hunk 2]
```diff
@@ -25,6 +25,8 @@ jobs:
 
       - name: Gradle Setup
         uses: gradle/gradle-build-action@v2
+        with:
+            cache-disabled: true
```
Comment (vitorhugods) [inline .github/workflows/gradle-android-tests.yml:29]: Can we revert this now?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt [hunk 3]
```diff
@@ -14,7 +14,7 @@ import com.wire.kalium.logic.data.user.UserId
 import com.wire.kalium.logic.functional.Either
 import com.wire.kalium.logic.functional.suspending
 import com.wire.kalium.logic.wrapCryptoRequest
-import com.wire.kalium.cryptography.UserId as CryptoUserId
+import com.wire.kalium.cryptography.PlainUserId as CryptoUserId
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt:17]: I think we can remove the alias now that it was renamed to `PlainUserId`

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/ConversationEventReceiver.kt [hunk 4]
```diff
@@ -3,7 +3,7 @@ package com.wire.kalium.logic.sync
 import com.wire.kalium.cryptography.CryptoClientId
 import com.wire.kalium.cryptography.CryptoSessionId
 import com.wire.kalium.cryptography.ProteusClient
-import com.wire.kalium.cryptography.UserId
+import com.wire.kalium.cryptography.PlainUserId
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/ConversationEventReceiver.kt:6]: I think we can remove the alias now that it was renamed to `PlainUserId`

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreatorTest.kt [hunk 5]
```diff
@@ -31,7 +31,7 @@ import kotlin.test.Test
 import kotlin.test.assertEquals
 import kotlin.test.assertIs
 import kotlin.test.assertTrue
-import com.wire.kalium.cryptography.UserId as CryptoUserId
+import com.wire.kalium.cryptography.PlainUserId as CryptoUserId
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreatorTest.kt:34]: Here too

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisherTest.kt [hunk 6]
```diff
@@ -4,7 +4,7 @@ import com.wire.kalium.cryptography.CryptoClientId
 import com.wire.kalium.cryptography.CryptoSessionId
 import com.wire.kalium.cryptography.PreKeyCrypto
 import com.wire.kalium.cryptography.ProteusClient
-import com.wire.kalium.cryptography.UserId
+import com.wire.kalium.cryptography.PlainUserId
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisherTest.kt:7]: And here as well

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreator.kt [hunk 7]
```diff
@@ -15,7 +15,7 @@ import com.wire.kalium.logic.data.message.RecipientEntry
 import com.wire.kalium.logic.functional.Either
 import com.wire.kalium.logic.functional.suspending
 import com.wire.kalium.logic.wrapCryptoRequest
-import com.wire.kalium.cryptography.UserId as CryptoUserId
+import com.wire.kalium.cryptography.PlainUserId as CryptoUserId
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreator.kt:18]: Here

**Review by vitorhugods — APPROVED** [review]
👨🏼‍🍳👌🏼

File: .github/workflows/gradle-android-tests.yml (hunk 2 above)
Comment (typfel) [reply .github/workflows/gradle-android-tests.yml]: Reverted

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/ConversationEventReceiver.kt (hunk 4 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/ConversationEventReceiver.kt:6]: There's no alias here.

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisherTest.kt (hunk 6 above)
Comment (typfel) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisherTest.kt:7]: There's no alias here.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt (hunk 3 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SessionEstablisher.kt]: Done

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreatorTest.kt (hunk 5 above)
Comment (typfel) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/MessageEnvelopeCreatorTest.kt]: Done

**Review by MohamadJaara — APPROVED** [review]
